pkg/storage/vault.go

```diff
@@ -0,0 +1,231 @@
+package storage
+
+import (
+	"bytes"
+	"encoding/json"
+	"fmt"
+	"html/template"
+	"io/ioutil"
+	"os"
+	"path"
+	"strings"
+
+	"github.com/dollarshaveclub/psst/pkg/directory"
+	"github.com/hashicorp/vault/api"
+	"github.com/pkg/errors"
+)
+
+const (
+	keyPrefix       = "/secret/psst"
+	vaultSecretName = "secret"
+
+	filePerms = 0750
+)
+
+// PolicyTemplates are used with Vault to give users permissions
+type PolicyTemplates struct {
+	GeneralPolicyTemplate string
+	PolicyTemplate        string
+}
+
+type ghUserPolicy struct {
+	Value string `json:"value"`
+}
+type space struct {
+	Path string
+}
+
+var (
+	policies = map[string]PolicyTemplates{
+		"github": PolicyTemplates{
+			GeneralPolicyTemplate: `# Allows all users to write secrets to other users
+path "{{.Path}}/*" {
+	capabilities = ["create", "update"]
+}
+`,
+			PolicyTemplate: `# Allows a user to read secrets from personal drop keyspace
+path "{{.Path}}/*" {
+	capabilities = ["read", "list", "delete"]
+}
+`,
+		},
+	}
+)
+
+// VaultStore stores a Vault client
+type VaultStore struct {
+	*api.Client
+}
+
+// NewVault will connect to a Vault server using VAULT_ADDR and VAUL_TOKEN variables
+func NewVault() (*VaultStore, error) {
+	client, err := api.NewClient(api.DefaultConfig())
+	if err != nil {
+		return &VaultStore{}, fmt.Errorf("unable to get Vault client: %+v", err)
+	}
+	return &VaultStore{client}, nil
+}
+
+// Get will return the stored secret at a given path
+func (v *VaultStore) Get(path string) (string, error) {
+	secret, err := v.Client.Logical().Read(path)
+	if err != nil {
+		return "", fmt.Errorf("unable to read secret from vault: %+v", err)
+	}
+	if secret == nil {
+		return "", errors.New("no secret found")
+	}
+	if data, ok := secret.Data[vaultSecretName]; ok {
+		return data.(string), nil
+	}
+	return "", errors.New("improperly formatted secret")
+}
+
+// Write will write the provided secret to the given user
+func (v *VaultStore) Write(filename, name string, targets map[string]interface{}) error {
+	buf, err := ioutil.ReadFile(filename)
```

Comment:
Consider doing a sanity check on the file size before trying to read the entire thing into memory.